src/config/graph.rs

```diff
@@ -0,0 +1,559 @@
+                    | (Node::Source { ty: ty1 }, Node::Transform { in_ty: ty2, .. })
+                    | (Node::Transform { out_ty: ty1, .. }, Node::Transform { in_ty: ty2, .. })
+                    | (Node::Transform { out_ty: ty1, .. }, Node::Sink { ty: ty2, .. }) => {
+                        if ty1 != ty2 && ty1 != DataType::Any && ty2 != DataType::Any {
+                            errors.push(format!(
+                                "Data type mismatch between {} ({:?}) and {} ({:?})",
+                                x, ty1, y, ty2
+                            ));
+                        }
+                    }
+                    (Node::Sink { .. }, _) | (_, Node::Source { .. }) => unreachable!(),
+                }
+            }
+        }
+
+        if errors.is_empty() {
+            Ok(())
+        } else {
+            errors.sort();
+            errors.dedup();
+            Err(errors)
+        }
+    }
+
+    pub fn valid_inputs(&self) -> HashSet<OutputId> {
+        self.nodes
+            .iter()
+            .flat_map(|(key, node)| match node {
+                Node::Sink { .. } => vec![],
+                Node::Source { .. } => vec![key.clone().into()],
+                Node::Transform { named_outputs, .. } => {
+                    let mut outputs = vec![key.clone().into()];
+                    outputs.extend(
+                        named_outputs
+                            .clone()
+                            .into_iter()
+                            .map(|n| OutputId::from((key, n))),
+                    );
+                    outputs
+                }
+            })
+            .collect()
+    }
+
+    /// Produce a map of output IDs for the current set of nodes in the graph, keyed by their string
+    /// representation. Returns errors for any nodes that have the same string representation,
+    /// making input specifications ambiguous.
+    ///
+    /// When we get a dotted path in the `inputs` section of a user's config, we need to determine
+    /// which of a few things that represents:
+    ///
+    ///   1. A component that's part of an expanded macro (e.g. `route.branch`)
+    ///   2. A component within a pipeline (e.g. `pipeline.name`)
+    ///   3. A named output of a branching transform (e.g. `name.errors`)
+    ///
+    /// A naive way to do that is to compare the string representation of all valid inputs to the
+    /// provided string and pick the one that matches. This works better if you can assume that there
+    /// are no conflicting string representations, so this function reports any ambiguity as an
+    /// error when creating the lookup map.
+    pub fn input_map(&self) -> Result<HashMap<String, OutputId>, Vec<String>> {
+        let all_outputs = self.valid_inputs();
+        let mut mapped: HashMap<String, OutputId> = HashMap::new();
+        for id in all_outputs {
+            if let Some(_other) = mapped.insert(id.to_string(), id) {
+                panic!("double");
```

Comment:
This is a little obscure for a panic message, even though it should never happen.